### PR TITLE
feat(VCalendar): allow outside item to drop on interval

### PR DESCRIPTION
**Pull request by YipingRuan**
## Description

Drop an outside item and add event

## How Has This Been Tested?

visually 

## Markup:

<details>

```vue
// Paste your FULL Playground.vue here
<template>
  <div id="app">
    <v-row class="fill-height">
      <v-col>
        <v-card
          max-width="244"
          draggable
          elevation="2"
          outlined
        >
          <v-list-item three-line>
            <v-list-item-content>
              <div class="overline mb-4"></div>
              <v-list-item-title class="headline mb-1">
                Drag me
              </v-list-item-title>
              <v-list-item-subtitle>to calendar</v-list-item-subtitle>
            </v-list-item-content>

            <v-list-item-avatar
              tile
              size="80"
              color="grey"
            ></v-list-item-avatar>
          </v-list-item>
        </v-card>
        <v-sheet height="600">
          <v-calendar
            ref="calendar"
            v-model="value"
            color="primary"
            type="4day"
            :events="events"
            :event-color="getEventColor"
            :event-ripple="false"
            first-time="08:00"
            first-interval="8"
            last-time="20:00"
            interval-count="12"
            @change="getEvents"
            @mousedown:event="startDrag"
            @mousedown:time="startTime"
            @mousemove:time="mouseMove"
            @mouseup:time="endDrag"
            @mouseleave.native="cancelDrag"
            @drop:event="intervalDrop"
          >
            <template #event="{ event, timed, eventSummary }">
              <div
                class="v-event-draggable"
                v-html="eventSummary()"
              ></div>
              <div
                v-if="timed"
                class="v-event-drag-bottom"
                @mousedown.stop="extendBottom(event)"
              ></div>
            </template>
          </v-calendar>
        </v-sheet>
      </v-col>
    </v-row>
  </div>

</template>

<script>
  export default {
    data: () => ({
      value: '',
      events: [],
      colors: ['#2196F3', '#3F51B5', '#673AB7', '#00BCD4', '#4CAF50', '#FF9800', '#757575'],
      names: ['Meeting', 'Holiday', 'PTO', 'Travel', 'Event', 'Birthday', 'Conference', 'Party'],
      dragEvent: null,
      dragStart: null,
      createEvent: null,
      createStart: null,
      extendOriginal: null,
    }),
    methods: {
      intervalDrop ({ event, interval }) {
        console.log(event, interval)
      },
      startDrag ({ event, timed }) {
        if (event && timed) {
          this.dragEvent = event
          this.dragTime = null
          this.extendOriginal = null
        }
      },
      startTime (tms) {
        const mouse = this.toTime(tms)

        if (this.dragEvent && this.dragTime === null) {
          const start = this.dragEvent.start

          this.dragTime = mouse - start
        } else {
          this.createStart = this.roundTime(mouse)
          this.createEvent = {
            name: `Event #${this.events.length}`,
            color: this.rndElement(this.colors),
            start: this.createStart,
            end: this.createStart,
            timed: true,
          }

          this.events.push(this.createEvent)
        }
      },
      extendBottom (event) {
        this.createEvent = event
        this.createStart = event.start
        this.extendOriginal = event.end
      },
      mouseMove (tms) {
        const mouse = this.toTime(tms)

        if (this.dragEvent && this.dragTime !== null) {
          const start = this.dragEvent.start
          const end = this.dragEvent.end
          const duration = end - start
          const newStartTime = mouse - this.dragTime
          const newStart = this.roundTime(newStartTime)
          const newEnd = newStart + duration

          this.dragEvent.start = newStart
          this.dragEvent.end = newEnd
        } else if (this.createEvent && this.createStart !== null) {
          const mouseRounded = this.roundTime(mouse, false)
          const min = Math.min(mouseRounded, this.createStart)
          const max = Math.max(mouseRounded, this.createStart)

          this.createEvent.start = min
          this.createEvent.end = max
        }
      },
      endDrag () {
        this.dragTime = null
        this.dragEvent = null
        this.createEvent = null
        this.createStart = null
        this.extendOriginal = null
      },
      cancelDrag () {
        if (this.createEvent) {
          if (this.extendOriginal) {
            this.createEvent.end = this.extendOriginal
          } else {
            const i = this.events.indexOf(this.createEvent)
            if (i !== -1) {
              this.events.splice(i, 1)
            }
          }
        }

        this.createEvent = null
        this.createStart = null
        this.dragTime = null
        this.dragEvent = null
      },
      roundTime (time, down = true) {
        const roundTo = 15 // minutes
        const roundDownTime = roundTo * 60 * 1000

        return down
          ? time - time % roundDownTime
          : time + (roundDownTime - (time % roundDownTime))
      },
      toTime (tms) {
        return new Date(tms.year, tms.month - 1, tms.day, tms.hour, tms.minute).getTime()
      },
      getEventColor (event) {
        const rgb = parseInt(event.color.substring(1), 16)
        const r = (rgb >> 16) & 0xFF
        const g = (rgb >> 8) & 0xFF
        const b = (rgb >> 0) & 0xFF

        return event === this.dragEvent
          ? `rgba(${r}, ${g}, ${b}, 0.7)`
          : event === this.createEvent
            ? `rgba(${r}, ${g}, ${b}, 0.7)`
            : event.color
      },
      getEvents ({ start, end }) {
        const events = []

        const min = new Date(`${start.date}T00:00:00`).getTime()
        const max = new Date(`${end.date}T23:59:59`).getTime()
        const days = (max - min) / 86400000
        const eventCount = this.rnd(days, days + 20)

        for (let i = 0; i < eventCount; i++) {
          const timed = this.rnd(0, 3) !== 0
          const firstTimestamp = this.rnd(min, max)
          const secondTimestamp = this.rnd(2, timed ? 8 : 288) * 900000
          const start = firstTimestamp - (firstTimestamp % 900000)
          const end = start + secondTimestamp

          events.push({
            name: this.rndElement(this.names),
            color: this.rndElement(this.colors),
            start,
            end,
            timed,
          })
        }

        this.events = events
      },
      rnd (a, b) {
        return Math.floor((b - a + 1) * Math.random()) + a
      },
      rndElement (arr) {
        return arr[this.rnd(0, arr.length - 1)]
      },
    },
  }
</script>

```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

**Comment by YipingRuan** [comment]
drag

**Comment by manuelmeister** [comment]
I think if v-calendar-daily has dragover and drop support to put something into the calendar, we should also have the drag event on the individual "calendar events".

**Comment by YipingRuan** [comment]
> I think if v-calendar-daily has dragover and drop support to put something into the calendar, we should also have the drag event on the individual "calendar events".

Do you mean this? https://vuetifyjs.com/en/components/calendars/#drag-and-drop

**Comment by manuelmeister** [comment]
@YipingRuan no those are only `@mousedown` & `@mouseup` and not `@dragstart`, …

**Comment by YipingRuan** [comment]
@johnleider   Should dragstart added at "genEvent" or leave it to the event slot (user)?

**Comment by manuelmeister** [comment]
@YipingRuan I think it would be confusing if it supported drop but not drag. Don`t you think so?

In the event slot, you cannot override the element where the class `.v-event-timed` is set. That means you can`t drag the element itself, but only an element that would be in it.

_(But I also believe that this is a significant design flaw with the `v-calendar-daily` component anyway)._

**Comment by YipingRuan** [comment]
@manuelmeister 
Understand your concern.  Yes I can certainly add "draggable" and emit the event.
image

What to do when the user only want some of the events are can be moved, not all (based on some logic, say if the event title contains some string)?

File: packages/vuetify/src/components/VCalendar/VCalendarDaily.ts [hunk 1]
```diff
@@ -192,6 +192,10 @@ export default CalendarWithIntervals.extend({
           height,
           ...styler(interval),
         },
+        on: {
+          dragover: (event: DragEvent) => event.preventDefault(),
+          drop: (event: DragEvent) => this.$emit('drop:event', { event, interval }),
```
Comment (KaelWD) [inline packages/vuetify/src/components/VCalendar/VCalendarDaily.ts:197]: - Wouldn't this be `drop:time`?
- Why not just add all the drag events to `getDefaultMouseEventHandlers`?

**Comment by johnleider** [comment]
This Pull Request is going being closed due to inactivity.

**Comment by YipingRuan** [comment]
> This Pull Request is going being closed due to inactivity.

I no longer have permission to access this branch..

**Comment by johnleider** [comment]
Okay. Closing this, open a new PR from a fork.

Thank you

**Pull request closed without merging**